modules/ensemble/lib/action/take_screenshot.dart

```diff
@@ -0,0 +1,111 @@
+import 'package:ensemble/screen_controller.dart';
+import 'package:ensemble/framework/scope.dart';
+import 'package:ensemble/util/utils.dart';
+import 'package:flutter/material.dart';
+import 'package:screenshot/screenshot.dart';
+
+class TakeScreenshotAction extends EnsembleAction {
+  TakeScreenshotAction({
+    super.initiator,
+    required this.widgetId,
+    this.pixelRatio,
+    this.onSuccess,
+    this.onFailure,
+  });
+
+  final dynamic widgetId;
+  final double? pixelRatio;
+  final EnsembleAction? onSuccess;
+  final EnsembleAction? onFailure;
+
+  factory TakeScreenshotAction.fromYaml({Map? payload}) {
+    if (payload == null || payload['widgetId'] == null) {
+      throw LanguageError(
+          "${ActionType.takeScreenshot.name} requires 'widgetId'");
+    }
+    return TakeScreenshotAction(
+      widgetId: payload['widgetId'],
+      pixelRatio: Utils.optionalDouble(payload['pixelRatio']),
+      onSuccess: payload['onSuccess'] != null
+          ? EnsembleAction.from(payload['onSuccess'])
+          : null,
+      onFailure: payload['onFailure'] != null
+          ? EnsembleAction.from(payload['onFailure'])
+          : null,
+    );
+  }
+
+  @override
+  Future<void> execute(BuildContext context, ScopeManager scopeManager) async {
+    final screenshotController = ScreenshotController();
+
+    try {
+      final resolvedWidget = scopeManager.dataContext.eval(widgetId);
+      if (resolvedWidget == null) {
+        throw LanguageError("Widget not found: '$widgetId'");
+      }
+
+      final widget = Screenshot(
+        controller: screenshotController,
+        child: DataScopeWidget(
+          scopeManager: scopeManager,
+          child: resolvedWidget,
+        ),
+      );
+
+      final Uint8List? capturedImage =
+          await screenshotController.captureFromLongWidget(
+        InheritedTheme.captureAll(
+          context,
+          Material(
+            type: MaterialType.transparency,
+            child: widget,
+          ),
+        ),
+        delay: const Duration(milliseconds: 1000),
+        pixelRatio: pixelRatio ?? MediaQuery.of(context).devicePixelRatio,
+        context: context,
+      );
+
+      if (capturedImage == null) {
+        throw LanguageError("Failed to capture screenshot");
+      }
+
+      final dimensions = await _getImageDimensions(capturedImage);
+
+      if (onSuccess != null) {
+        await ScreenController().executeAction(
+          context,
+          onSuccess!,
+          event: EnsembleEvent(initiator, data: {
+            'imageData': capturedImage,
+            'size': capturedImage.length,
+            'dimensions': dimensions,
+          }),
+        );
+      }
+    } catch (e) {
+      if (onFailure != null) {
+        await ScreenController().executeAction(
+          context,
+          onFailure!,
+          event: EnsembleEvent(initiator, data: {'error': e.toString()}),
+        );
+      }
+      rethrow;
```

Comment:
have you actually tested it? rethrowing the error may cause issues. TEST the error case please